src/complex_fluids/CFINSForcing.cpp

```diff
@@ -674,40 +690,49 @@ CFINSForcing::squareMatrix(const int data_idx,
                            const double /*data_time*/,
                            const bool initial_time,
                            const int coarsest_ln,
-                           const int finest_ln)
+                           const int finest_ln,
+                           const bool extended_box)
 {
     for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
     {
         Pointer<PatchLevel<NDIM> > level = hierarchy->getPatchLevel(ln);
         for (PatchLevel<NDIM>::Iterator p(level); p; p++)
         {
             Pointer<Patch<NDIM> > patch = level->getPatch(p());
-            const Box<NDIM>& box = patch->getBox();
-            const Pointer<PatchGeometry<NDIM> > p_geom = patch->getPatchGeometry();
             if (initial_time) return;
-            Pointer<CellData<NDIM, double> > half_data = patch->getPatchData(d_W_cc_idx);
-            Pointer<CellData<NDIM, double> > tens_data = patch->getPatchData(data_idx);
+            Pointer<CellData<NDIM, double> > data = patch->getPatchData(data_idx);
+            const Box<NDIM>& box = extended_box ? data->getGhostBox() : patch->getBox();
+            const Pointer<PatchGeometry<NDIM> > p_geom = patch->getPatchGeometry();
+
             for (CellIterator<NDIM> it(box); it; it++)
             {
                 CellIndex<NDIM> i = *it;
 #if (NDIM == 2)
-                (*tens_data)(i, 0) = (*half_data)(i, 0) * (*half_data)(i, 0) + (*half_data)(i, 2) * (*half_data)(i, 2);
-                (*tens_data)(i, 1) = (*half_data)(i, 2) * (*half_data)(i, 2) + (*half_data)(i, 1) * (*half_data)(i, 1);
-                (*tens_data)(i, 2) = (*half_data)(i, 0) * (*half_data)(i, 2) + (*half_data)(i, 1) * (*half_data)(i, 2);
+                Eigen::Matrix2d mat;
+                mat(0, 0) = (*data)(i, 0);
```

Comment:
Is the data ordering amenable to using some loops here instead of hard-coding indexing?

Reply:
It's not pretty, but the best I can think of is something like

```
for (int i = 0; i < NDIM; ++i)
{
    for (int j = 0; j <= i; ++j)
    {
        mat(i,j) = mat(j,i) = i == j ? (*data)(i,0) : (*data)(3*NDIM-3 - i - j);
    }
}
```

Would this be preferable?

Reply:
This is Voigt notation, right? Can you make a helper function for Voigt notation?

Reply:
Sure. Something like `voigtToTensor` and `tensorToVoigt`?

Reply:
Sure --- if it is static member function `voigt_to_tensor` or `voigt_to_matrix` sounds good.

Reply:
Or `voigt_to_matrix_index` or `voigt_to_tensor_idx`.